experimental/Schemes/SpecOpen.jl

```diff
@@ -0,0 +1,694 @@
+  return SpecOpenMor(U, SpecOpen(Y), maps)
+end
+
+function maximal_extension(
+    X::Spec{BRT, BRET, RT, RET, MST}, 
+    Y::Spec{BRT, BRET, RT, RET, MST}, 
+    f::Vector
+  ) where {BRT, BRET, RT, RET, MST}
+  return maximal_extension(X, Y, FractionField(base_ring(OO(X))).(f))
+end
+
+function restriction(
+    f::SpecOpenMor{BRT, BRET, RT, RET, MST1, MST2},
+    U::SpecOpen{BRT, BRET, RT, RET, MST1},
+    V::SpecOpen{BRT, BRET, RT, RET, MST2}
+  ) where {BRT, BRET, RT, RET, MST1, MST2}
+  inc = SpecOpenMor(U, domain(f), [SpecMor(W, ambient(domain(f)), gens(localized_ring(OO(W)))) for W in patches(U)])
+  help_map = compose(inc, f)
+  return SpecOpenMor(U, V, maps_on_patches(help_map))
+end
+
+identity_map(U::SpecOpen) = SpecOpenMor(U, U, [SpecMor(V, ambient(U), gens(localized_ring(OO(V)))) for V in patches(U)])
+
+function ==(f::T, g::T) where {T<:SpecOpenMor} 
+  domain(f) == domain(g) || return false
+  codomain(f) == codomain(g) || return false
+  Y = ambient(codomain(f))
+  m = length(patches(domain(f)))
+  n = length(patches(domain(g)))
+  for i in 1:m
+    for j in 1:n
+      restrict(f[i], intersect(domain(f)[i], domain(g)[i]), Y) == 
+      restrict(g[i], intersect(domain(f)[i], domain(g)[i]), Y) || return false
+    end
+  end
+  return true
+end
+
+function preimage(f::SpecOpenMor{BRT, BRET, RT, RET, MST1, MST2},
+    Z::Spec{BRT, BRET, RT, RET, MST2}
+  ) where {BRT, BRET, RT, RET, MST1, MST2}
+  U = domain(f) 
+  X = ambient(U)
+  n = length(patches(U))
+  W = localized_ring(OO(X))
+  I = ideal(W, one(W))
+  for i in 1:n 
+    I = intersect(I, localized_modulus(OO(closure(preimage(f[i], Z), X))))
+  end
+  fZbar = subscheme(X, I)
+  return SpecOpen(fZbar, gens(U))
+end
+
+function preimage(f::SpecOpenMor{BRT, BRET, RT, RET, MST1, MST2},
+    V::SpecOpen{BRT, BRET, RT, RET, MST2}
+  ) where {BRT, BRET, RT, RET, MST1, MST2}
+  U = domain(f)
+  X = ambient(U)
+  R = base_ring(OO(X))
+  I = ideal(R, one(R))
+  for i in 1:npatches(U)
+    I = intersect(I, saturated_ideal(ideal(OO(U[i]), pullback(f[i]).(gens(V)))))
+  end
+  return intersect(U, SpecOpen(X, gens(I)))
+end
+
+function is_non_zero_divisor(f::RET, U::SpecOpen{BRT, BRET, RT, RET}) where {BRT, BRET, RT, RET}
+  for V in patches(U)
+    zero_ideal = ideal(OO(V), [zero(OO(V))])
+    zero_ideal == quotient(zero_ideal, ideal(OO(V), [f])) || return false
+  end
+  return true
+end
+
+function find_non_zero_divisor(U::SpecOpen)
+  n = length(gens(U))
+  X = ambient(U)
+  kk = coefficient_ring(base_ring(OO(X)))
+  d = dot([rand(kk, 0:100) for i in 1:n], gens(U))
+  while !is_non_zero_divisor(d, U)
+    d = dot([rand(kk, 0:100) for i in 1:n], gens(U))
+  end
+  return d
+end
+
+function generic_fractions(f::SpecOpenMor)
+  U = domain(f)
+  X = ambient(U)
+  V = codomain(f)
+  Y = ambient(V)
+  d = find_non_zero_divisor(U)
+  V = hypersurface_complement(X, d)
+  result = fraction.([restrict(pullback(f, y), V) for y in gens(base_ring(OO(Y)))])
+end
+
+function is_dense(U::SpecOpen)
```

Comment:
`isdense`

Reply:
Nope. Decisions have been taken lately that require the underscore for everything in the future.

Reply:
Oh. Always?

Reply:
if this is the case this needs to be documented in
https://oscar-system.github.io/Oscar.jl/dev/DeveloperDocumentation/styleguide/

Reply:
I don't think we said the underscore are *required*; rather, before they were forbidden, now they are not.

I actually thought that we updated the docs... @thofma didn't you want to do that (I might misremember; unfortunately I couldn't find the relevant Oscar issue right now ...)

Reply:
Not yet. But we had the consensus that they should be required.